merk/src/tree/ops.rs

```diff
@@ -1,21 +1,56 @@
-use std::{collections::LinkedList, fmt};
+use std::{
+    collections::{BTreeSet, LinkedList},
+    fmt,
+};
 
 use anyhow::Result;
-use costs::{cost_return_on_error, CostContext, CostsExt, OperationCost};
+use costs::{
+    cost_return_on_error, cost_return_on_error_no_add,
+    storage_cost::{
+        key_value_cost::KeyValueStorageCost, removal::StorageRemovedBytes::BasicStorageRemoval,
+        StorageCost,
+    },
+    CostContext, CostsExt, OperationCost,
+};
+use integer_encoding::VarInt;
 use Op::*;
 
 use super::{Fetch, Link, Tree, Walker};
-use crate::Hash;
+use crate::{CryptoHash, HASH_LENGTH_U32};
 
 /// Type alias to add more sense to function signatures.
-type DeletedKeys = LinkedList<Vec<u8>>;
+type UpdatedRootKeyFrom = Option<Vec<u8>>;
+
+/// Type alias to add more sense to function signatures.
+type NewKeys = BTreeSet<Vec<u8>>;
+
+/// Type alias to add more sense to function signatures.
+type UpdatedKeys = BTreeSet<Vec<u8>>;
+
+/// Type alias to add more sense to function signatures.
+type DeletedKeys = LinkedList<(Vec<u8>, Option<KeyValueStorageCost>)>;
 
 /// An operation to be applied to a key in the store.
-#[derive(PartialEq, Eq)]
+#[derive(PartialEq, Clone, Eq)]
 pub enum Op {
     Put(Vec<u8>),
-    PutReference(Vec<u8>, Hash),
+    /// Combined references include the value in the node hash
+    /// because the value is independent of the reference hash
+    /// In GroveDB this is used for references
+    PutCombinedReference(Vec<u8>, CryptoHash),
+    /// Layered references include the value in the node hash
+    /// because the value is independent of the reference hash
+    /// In GroveDB this is used for trees
+    /// A layered reference does not pay for the tree's value,
+    /// instead providing a cost for the value
+    PutLayeredReference(Vec<u8>, u32, CryptoHash),
+    /// Replacing a layered reference is slightly more efficient
+    /// than putting it as the replace will not modify the size
+    /// hence there is no need to calculate a difference in
+    /// costs.
+    ReplaceLayeredReference(Vec<u8>, u32, CryptoHash),
     Delete,
+    DeleteLayered,
```

Comment:
We should probably add a comment describing DeleteLayered

Reply:
Done